File: files/jvm/src/main/java/com/swoval/files/QuickFile.java
package com.swoval.files;

import static com.swoval.files.QuickListerImpl.DIRECTORY;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * Represents a file that will be returned by {@link QuickList#list}. Provides fast {@link
 * QuickFile#isDirectory} and {@link QuickFile#isFile} methods that should not call stat (or the
 * non-POSIX equivalent) on the underlying file. Can be converted to a {@link java.io.File} or
 * {@link java.nio.file.Path} with {@link QuickFile#toFile} and {@link QuickFile#toPath}.
 */
public interface QuickFile {
  /**
   * Returns the fully resolved file name
   *
   * @return the fully resolved file name
   */
  String getFileName();

  /**
   * Returns true if this was a directory at the time time of listing. This may become inconsistent
   * if the QuickFile is cached
   *
   * @return true when the QuickFile is a directory
   */
  boolean isDirectory();
  /**
   * Returns true if this was a regular file at the time time of listing. This may become
   * inconsistent if the QuickFile is cached
   *
   * @return true when the QuickFile is a file
   */
  boolean isFile();

  /**
   * Returns true if this was a symbolic link at the time time of listing. This may become
   * inconsistent if the QuickFile is cached
   *
   * @return true when the QuickFile is a symbolic link
   */
  boolean isSymbolicLink();

  /**
   * Returns an instance of {@link java.io.File} with fast implementations of {@link
   * java.io.File#isFile} and {@link File#isDirectory}. The {@link java.io.File#isFile} and {@link
   * java.io.File#isDirectory} methods return the cached values returned by the native file result
   * returned by readdir (posix) or FindNextFile (windows). This makes the result generically
   * unsuitable for caching. Instead, use {@link QuickFile#toFile}.
   *
   * @return An instance of {@link java.io.File}
   */
  File asFile();

  /**
   * Returns an instance of {@link java.io.File}. The instance should not override {@link
   * java.io.File#isDirectory} or {@link java.io.File#isFile} which makes it safe to persist.
   *
   * @return an instance of {@link java.io.File}
   */
  File toFile();

  /**
   * Returns an instance of {@link java.nio.file.Path}.
   *
   * @return an instance of {@link java.nio.file.Path}
   */
  Path toPath();
}

class QuickFileImpl extends File implements QuickFile {
  private final int kind;

  QuickFileImpl(final String name, final int kind) {
    super(name);
    this.kind = kind;
  }

  @Override
  public String getFileName() {
    return super.toString();
  }

  @Override
  public boolean isDirectory() {
    return is(QuickListerImpl.UNKNOWN) ? super.isDirectory() : is(DIRECTORY);
  }

  @Override
  public boolean isFile() {
    return is(QuickListerImpl.UNKNOWN) ? super.isFile() : is(QuickListerImpl.FILE);
  }

  @Override
  public boolean isSymbolicLink() {
    return is(QuickListerImpl.UNKNOWN) ? Files.isSymbolicLink(toPath()) : is(QuickListerImpl.LINK);
  }

  @Override
  public File asFile() {
    return this;
  }

  @Override
  public File toFile() {
    return new File(getFileName());
  }

  @Override
  public String toString() {
    return "QuickFile(" + getFileName() + ")";
  }

  @Override
  public boolean equals(Object other) {
    return other instanceof QuickFile
        && this.getFileName().equals(((QuickFile) other).getFileName());
  }

  @Override
  public int hashCode() {
    return toString().hashCode();
  }

  private boolean is(final int kind) {
    return (this.kind & kind) != 0;
  }
}

File: files/js/src/main/scala/com/swoval/files/QuickFile.scala
package com.swoval.files

import com.swoval.files.QuickListerImpl.DIRECTORY
import java.io.File
import java.nio.file.Files
import java.nio.file.Path

abstract class FileWithFileType(name: String) extends File(name) with FileType

/**
 * Represents a file that will be returned by [[QuickList.list]]. Provides fast [[QuickFile.isDirectory]] and [[QuickFile.isFile]] methods that should not call stat (or the
 * non-POSIX equivalent) on the * underlying file. Can be converted to a [[java.io.File]] or
 * [[java.nio.file.Path]] with [[QuickFile.toFile]] and [[QuickFile.toPath]].
 */
trait QuickFile {

  /**
   * Returns the fully resolved file name
   *
   * @return the fully resolved file name
   */
  def getFileName(): String

  /**
   * Returns true if this was a directory at the time time of listing. This may become inconsistent
   * if the QuickFile is cached
   *
   * @return true when the QuickFile is a directory
   */
  def isDirectory(): Boolean

  /**
   * Returns true if this was a regular file at the time time of listing. This may become
   * inconsistent if the QuickFile is cached
   *
   * @return true when the QuickFile is a file
   */
  def isFile(): Boolean

  /**
   * Returns true if this was a symbolic link at the time time of listing. This may become
   * inconsistent if the QuickFile is cached
   *
   * @return true when the QuickFile is a symbolic link
   */
  def isSymbolicLink(): Boolean

  /**
   * Returns an instance of [[FileWithFileType]]. Typically the implementation of [[QuickFile]] while extend [[FileWithFileType]]. This method will then just cast the instance
   * to [[java.io.File]]. Because the [[QuickFile.isDirectory]] and [[QuickFile.isFile]]
   * methods will generally cache the value of the native file result returned by readdir (posix) or
   * FindNextFile (windows) and use this value to compute [[QuickFile.isDirectory]] and [[QuickFile.isFile]], the returned [[FileWithFileType]] is generally unsuitable to be used as
   * a persistent value. Instead, use [[QuickFile.toFile]].
   *
   * @return An instance of FileWithFileType. This may just be a cast.
   */
  def asFile(): FileWithFileType

  /**
   * Returns an instance of [[java.io.File]]. The instance should not override [[java.io.File.isDirectory]] or [[java.io.File.isFile]] which makes it safe to persist.
   *
   * @return an instance of [[java.io.File]]
   */
  def toFile(): File

  /**
   * Returns an instance of [[java.nio.file.Path]].
   *
   * @return an instance of [[java.nio.file.Path]]
   */
  def toPath(): Path

}

private[files] class QuickFileImpl(name: String, private val kind: Int)
    extends FileWithFileType(name)
    with QuickFile {

  override def getFileName(): String = super.toString

  override def isDirectory(): Boolean =
    if (is(QuickListerImpl.UNKNOWN)) super.isDirectory else is(DIRECTORY)

  override def isFile(): Boolean =
    if (is(QuickListerImpl.UNKNOWN)) super.isFile else is(QuickListerImpl.FILE)

  override def isSymbolicLink(): Boolean =
    if (is(QuickListerImpl.UNKNOWN)) Files.isSymbolicLink(toPath())
    else is(QuickListerImpl.LINK)

  override def asFile(): FileWithFileType = this

  override def toFile(): File = new File(getFileName)

  override def toString(): String = "QuickFile(" + getFileName + ")"

  override def equals(other: Any): Boolean = other match {
    case other: QuickFile => this.getFileName == other.getFileName
    case _                => false

  }

  override def hashCode(): Int = toString.hashCode

  private def is(kind: Int): Boolean = (this.kind & kind) != 0

}
